### DISCUSSION TITLE
Subquery

### DISCUSSION THREAD

**Discussion by AdaLam7** [bytepayment/bytepay#20] [Ideas]
<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-ba483dd6-7fff-99ba-8537-bd7577209636"><h3 dir="ltr" style="line-height:1.5;background-color:#ffffff;margin-top:18pt;margin-bottom:12pt;"><span style="font-size:16.5pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Milestone 3 — </span><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Opensource product NFT</span></h3><ul style="margin-top:0;margin-bottom:0;padding-inline-start:48px;"><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Estimated Duration: 2 weeks</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">FTE: 3</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:12pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Costs: 15,000 USD</span></p></li></ul><p dir="ltr" style="line-height:1.38;margin-left: 36pt;background-color:#ffffff;margin-top:3pt;margin-bottom:12pt;">&nbsp;</p><div dir="ltr" style="margin-left:0pt;" align="left">

Number | Deliverable | Specification
-- | -- | --
0a. | License | Apache 2.0
0b. | Documentation | We will provide docs about how to create a task, how to tigger the payment, and how to withdraw
0c. | Testing Guide | We will provide uni test for task,transfer,informal,withdraw modules(70% cover), and UI test report
0d. | Docker | We will provide Dockerfile and docker image to run the website in one command line, and you can to test those modules
01. | Mint NFT | Using Opensrouce Product like bin apk docker image to mint NFT Mange the NFT price,circulation and other metadata.
02. | NFT market | Marketplace to show all opensource NFTs, and search NFT what you want.
03. | NFT details and buy NFT | Details to show the NFT metadata like price, owner, version. Anyone can buy what opensource product they need.

</div></b>
</body>
</html>

<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-ed4437e9-7fff-89c5-847b-038477575e9a"><h3 dir="ltr" style="line-height:1.5;background-color:#ffffff;margin-top:18pt;margin-bottom:0pt;padding:0pt 0pt 12pt 0pt;"><span style="font-size:16.5pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Milestone 2 — Transfer module, task module, informal, withdraw module</span></h3><br /><ul style="margin-top:0;margin-bottom:0;padding-inline-start:48px;"><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Estimated Duration: 2 weeks</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">FTE: 3</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:12pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Costs: 15,000 USD</span></p></li></ul><p dir="ltr" style="line-height:1.38;margin-left: 36pt;background-color:#ffffff;margin-top:3pt;margin-bottom:12pt;">&nbsp;</p><div dir="ltr" style="margin-left:0pt;" align="left">

Number | Deliverable | Specification
-- | -- | --
0a. | License | Apache 2.0
0b. | Documentation | We will provide docs about how to create a task, how to tigger the payment, and how to withdraw
0c. | Testing Guide | We will provide uni test for task,transfer,informal,withdraw modules(70% cover), and UI test report
0d. | Docker | We will provide Dockerfile and docker image to run the website in one command line, and you can to test those modules
1.   | Task management | Create a paid task by comment an issue, it will trigger the create task event though the webhook, and webhook server will save the task and show it on our page, when developer complete the task, will tigger transfer module to pay the developer
2. | Tansfer module | Trigger a payment by comment an issue, like /pay Bob 10tokens, the tokens will transfer to developer platform account
3. | Withdraw module | Withdraw  tokens from our platform to developer own wallet, if developer bind it own address, payment will transfer to the account directly
4. | Informal | Developer will receive the event, tell him how to withdraw  tokens in our platform, robot will send Bob email and comment the issue

</div></b>
</body>
</html>

<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-78886983-7fff-e9fc-fe32-5507b685c0e0"><h3 dir="ltr" style="line-height:1.5;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;padding:6pt 0pt 12pt 0pt;">&nbsp;</h3><h3 dir="ltr" style="line-height:1.5;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;padding:6pt 0pt 12pt 0pt;"><span style="font-size:16.5pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Milestone 1 — User account management &amp; repo management &amp; mnemonic management</span></h3><br /><ul style="margin-top:0;margin-bottom:0;padding-inline-start:48px;"><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Estimated Duration: 2 weeks</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">FTE: 3</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:12pt;" role="presentation"><span style="font-size:12pt;font-family:Arial;color:#24292f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Costs: 15,000 USD</span></p></li></ul><p dir="ltr" style="line-height:1.38;margin-left: 36pt;background-color:#ffffff;margin-top:3pt;margin-bottom:12pt;">&nbsp;</p><div dir="ltr" style="margin-left:0pt;" align="left">

Number | Deliverable | Specification
-- | -- | --
0a. | License | Apache 2.0
0b. | Documentation | BytePay user guide, workflow, and what is BytePay.
0c. | Testing Guide | We will provide uni test for user management and recharge management modules(70% cover), the two module UI test report
0d. | Docker | We will provide Dockerfile and docker image to run the website in one command line
1. | User management, create an Moonbeam account for each developer | We will provide login function, you can login our website using github, and we will create an  account for each user by default
2. | Repo & webhook management | We will fetch your github repo list, so you can active the repo you want integrate with SubQuery , the webhook module will listen the pay event and tigger transform module to pay developer tokens
3. | Address binding | Develop using github issue comment to bind there  address. /bytepay bind [address]
4. | Recharge management | Recharge  tokens to your platform account
5. | Transfer EVM contract | We will provide an tested EVM contract, provide transfer limit, whitelist and transfer function.

</div></b>
</body>
</html>

**Comment by fanux** [comment]
Firstly, we would like to say thank you so much for taking the time to apply for the SubQuery Grants Programme.

We have reviewed your application and we require some further information to support us in determining the outcome.

Could you please update your milestone document with the below information? The attached guide will support you.
Please be aware that we have a maximum funding amount of $30,000 therefore your cost projection must be less than this.
Clearly outline where in your milestones you will be using SubQuery to index and query data - including what type of data and what you will use this for in your UI.



Clearly allocate the amount of funding requested in each milestone. For example in milestone 1 you have requested $15,000 - we require you to assign this to each resource and the number of hours they will be dedicating so we can assess the amount of effort. User stories under each milestone would be an advantage.
Let me know once the document has been updated.

Once you are able to update your application with the above details, we will reassess your application and be in touch with the outcome.

If you have any questions, please don’t hesitate to reach out.

**Comment by cocmolla** [comment]
We have 3 mongo collections
1. repos - Get from Github REST API(https://api.github.com/user/repos)
```json
{
  "_id": "string",
  "owner_id": "number",
  "repo_id":"number",
  "owner_name": "string",
  "repo_name": "string",
  "meta": {
    "id": "number",
    "node_id": "string",
    "name": "string",
    "full_name": "string",
    "private": boolean,
    "owner": "Github public user info",
    "html_url": "url string",
    "description": "description",
    "fork": true,
    "url": "https://api.github.com/repos/fanux/adidoks",
}
```
2. tasks - Define every task that open-source project author issued
```json
{
  "_id": "mongo object id",
  "issue_id": "number",
  "issue_number": "number",
  "title": "sting: issue title",
  "task_url": "url string",
  "repo_name": "repo full name",
  "repo_id":"repo string id",
  "repo_url": "string url",
  "author": "Github public user info",
  "pay": "number: price of current task",
  "describe": "string: Describe The Task",
  "createTime": "2022-01-18T07:55:08Z",
  "updataTime": "2022-01-18T07:55:08Z",
  "status": "finished",
  "developer": "Github public user info"
}
```
3. users - Based on github public user info and extend cryptocurrency information
```json
"user": "Github public userinfo",
"polka": "polka account info",
"near": "near account info",
"more...": "other cryptocurrency account info"
```

**Comment by sulnong** [comment]
Add more collections used by Bytepay

4. webhook payload - payload comes from github, we would save bytepay related comments
```json
  "action": "created, comment, delete, close, github issue and comment action",
  "issue": {
    "id": "issue id",
    "title": "issue title"
   "body": "issue body content"
  },
  "comment": {
   "body": "most import field, we check this comment and dispatch our logic"
  },
  "sender": "github public user info, who created or sended this issue or comment"
  "repository": "github public repository info, in which repso this issue and comment belong to"
```

5. nft - user can mint and sell open-source project as nft
```json
{
  "_id": "mongo object id",
  "title": "open-source project nft tilte",
  "price": "number",
   "unit": "unit of price, can be dolloar, dot, btc or any other cryptocurrency",
  "version": "project version",
  "project": "github project, eg. fanux/sealos",
  "total_supply": "number, total amounts in sell for this nft",
  "left_amount": "how many remain could be sell",
  "description": "a html rich text to describe this open-source project and nft",
  "chain_id": "nft unique id on bytepay chain",
  "owner": "who mint this nft",
  "owner_address": "13phH6rDBchj3mpMLTojff6osxHuupF1p4Gi87P5ZPbLBtmV",
  "id": "owner id, same as github user id",
  "classid": "0x00000001",
  "file_path": "we would backup project file, this field shows its location on our server",
  "file_hash": "0x236a7c5b2e3f95457afb90db6260f818a5d70d6f9804ec1718267c75307eb581",
  "file_cid": "cid on ipfs",
  "buyer": "buyers array， redundancy filed for nft orders collection",
  "status": 0,
  "created_time": "created time"
}
```

6. nft orders - save nft orders info, we saved a few `redundancy  fileds`
```json
{
  "_id": "mongo object id",
  "nft_id": "mongo object id in nft collection",
  "title": "open-source project id",
  "price": "number",
  "unit": "unit of price, can be dolloar, dot, btc or any other cryptocurrency"
  "version": "version name of project",
  "project": "github project, eg. fanux/sealos",
  "file_path": "we would backup project file, this field shows its location on our server",
  "file_hash": "0x236a7c5b2e3f95457afb90db6260f818a5d70d6f9804ec1718267c75307eb581",
  "file_cid": "cid on ipfs",
  "buyer_id": "github user id, who buys this nft",
  "buy_time": "time when this user bought"
}
```
